tests/raftstore/pd.rs

```diff
@@ -82,114 +99,357 @@ impl Cluster {
         Ok(self.stores.get(&store_id).unwrap().store.clone())
     }
 
-    fn get_region(&self, key: Vec<u8>) -> Result<metapb::Region> {
-        let (_, region) = self.regions
+    fn get_region(&self, key: Vec<u8>) -> Option<metapb::Region> {
+        self.regions
             .range::<Key, Key>(Excluded(&key), Unbounded)
             .next()
-            .unwrap();
-        Ok(region.clone())
+            .map(|(_, region)| region.clone())
     }
 
     fn get_region_by_id(&self, region_id: u64) -> Result<metapb::Region> {
         let key = self.region_id_keys.get(&region_id).unwrap();
-        if self.regions.contains_key(key) {
-            Ok(self.regions[key].clone())
+        Ok(self.regions.get(key).cloned().unwrap())
+    }
+
+    fn get_stores(&self) -> Vec<metapb::Store> {
+        self.stores.values().map(|s| s.store.clone()).collect()
+    }
+
+    fn add_region(&mut self, region: &metapb::Region) {
+        let end_key = enc_end_key(region);
+        assert!(self.regions.insert(end_key.clone(), region.clone()).is_none());
+        assert!(self.region_id_keys.insert(region.get_id(), end_key.clone()).is_none());
+    }
+
+    fn remove_region(&mut self, region: &metapb::Region) {
+        let end_key = enc_end_key(region);
+        assert!(self.regions.remove(&end_key).is_some());
+        assert!(self.region_id_keys.remove(&region.get_id()).is_some());
+    }
+
+    fn handle_heartbeat_version(&mut self, region: metapb::Region) -> Result<()> {
+        // For split, we should handle heartbeat carefully.
+        // E.g, for region 1 [a, c) -> 1 [a, b) + 2 [b, c).
+        // after split, region 1 and 2 will do heartbeat independently.
+        let start_key = enc_start_key(&region);
+        let end_key = enc_end_key(&region);
+        assert!(end_key > start_key);
+
+        let version = region.get_region_epoch().get_version();
+        let conf_ver = region.get_region_epoch().get_conf_ver();
+
+        let search_key = data_key(region.get_start_key());
+        let search_region = match self.get_region(search_key) {
+            None => {
+                // Find no range after start key, insert directly.
+                self.add_region(&region);
+                return Ok(());
+            }
+            Some(search_region) => search_region,
+        };
+
+        let search_start_key = enc_start_key(&search_region);
+        let search_end_key = enc_end_key(&search_region);
+
+        let search_version = search_region.get_region_epoch().get_version();
+        let search_conf_ver = search_region.get_region_epoch().get_conf_ver();
+
+        if start_key == search_start_key && end_key == search_end_key {
```

Comment:
Why not using `get_region_by_id` at L142? Then you will know they are the same Region.